feat: Enable CLI commands to run from anywhere in project (#3688)

FXschwartz opened the pull request:
Adds the ability to run the CLI commands (generate, create-migration, create-repair-migration) from anywhere in the repo by automatically locating the server package when running commands . The search respects repository boundaries (.git, melos.yaml) and handles monorepo/dart workspace structures with nested applications.

  - Auto-detects server directory from current location
  - Searches children, siblings, and upward through parents
  - Stops at repository/workspace boundaries
  - Prompts when multiple servers found
 
Fixes #3688 and #3530 

## Pre-launch Checklist

- [x] I read the [Contribute](https://docs.serverpod.dev/contribute) page and followed the process outlined there for submitting PRs.
- [x] This update contains only one single feature or bug fix and nothing else. (If you are submitting multiple fixes, please make multiple PRs.)
- [x] I read and followed the [Dart Style Guide](https://dart.dev/guides/language/effective-dart/style) and formatted the code with [dart format](https://dart.dev/tools/dart-format).
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`), and made sure that the documentation follows the same style as other Serverpod documentation. I checked spelling and grammar.
- [x] I added new tests to check the change I am making.
- [x] All existing and new tests are passing.
- [x] Any breaking changes are documented below.

If you need help, consider asking for advice on the [discussion board](https://github.com/serverpod/serverpod/discussions).

## Breaking changes
None

SandPod commented:
Hey @FXschwartz! Big thanks for taking this on!

Before examining your search algorithm, could you inspect the one we use for our Serverpod cloud CLI to ensure that it delivers a somewhat equal experience?

We have failed to port that code over to Serverpod yet, but now that you have area experience, perhaps you could make sure they make sense?

If they differ, feel free to argue for why your approach should be ported to that code instead of copying it.

The code for the Serverpod cloud cli search algorithm is available here: https://github.com/serverpod/serverpod_cloud_clients/blob/ed5f08c7e5da195a8ae4176f93ed1ea30f41fe4e/serverpod_cloud_cli/lib/util/scloud_config/file_finder.dart#L40

FXschwartz commented:
@SandPod Ok I've updated the search flow here to match the pattern used in `serverpod_cloud_cli` I did make a change to instead of calling it `FileFinder` here I used `DirectoryFinder` as it makes more sense to me in this context but feel free to have me change that.
- On a side note are there any plans to maybe move some of this file/directory search logic into a common place where we can share it? Not sure if there are any other places that search for specific files with logic as to how many levels it searches.

Additionally I've update all my test cases to use the `test_descriptor` package when creating temp directories.

And finally I added some test cases to validate the behavior in CI environments.

Take a look when you have some free time and let me know your thoughts or requested changes.

SandPod commented:
@FXschwartz Thanks for the update! I'll take a look ASAP.

We don't have a place to share this logic at the moment. So we will have to update it here for now, and then I'll have the team sort that out at some point.

SandPod commented on tools/serverpod_cli/lib/src/util/server_directory_finder.dart:
```diff
@@ -0,0 +1,314 @@
+  }
+
+  /// Prompts the user to select from multiple server directories.
+  static Future<Directory> _promptUserSelection(
```
Suggestion: I believe CLI tools have support for interactive prompts. 

More specifically a `select` prompt is available here: https://github.com/serverpod/cli_tools/blob/main/packages/cli_tools/lib/src/prompts/select.dart

Any reason for not using it?

SandPod commented on tools/serverpod_cli/lib/src/util/server_directory_finder.dart:
```diff
@@ -0,0 +1,314 @@
+  static Future<Directory> findOrPrompt({
+    Directory? startDir,
+    // Optional interactive property in case someone wanted to run these CLI commands in a CI/CD environment.
+    bool interactive = true,
```
Suggestion/Question: I can't see this option being configured from anywhere, and currently it is on by default. How would the user use this?

FXschwartz replied:
@SandPod Alright I've update the CLI commands (`generate`, `create-migration`, and `create-repair-migration`) to have this optional negatable flag. 

The priority is `CLI flag > CI detection > default (true)` with logic `final isInteractive = interactive ?? !ci.isCI;`

I also added the following e2e tests to cover this interactive logic.

  - Tests --no-interactive flag works
  - Tests CI environment auto-detection (CI=true)
  - Tests flag override behavior (--interactive overrides CI)
  - Tests multiple servers with --no-interactive → error message
  - Tests multiple servers with --interactive → shows prompt

I had to make multiple file changes to update the `GeneratorConfig.load()` call sites from positional to named parameters with the addition of the optional `bool? interactive` parameter.

Take a look and let me know what you'd like changed or any kind of feedback. 

Appreciate your time reviewing this PR!

SandPod replied:
Direction sounds good. I'll take a look at it during the weekend.

But could you make this a global flag instead, much like the verbose one?

I think the behaviour is that you don't care about what command you are running, but if there is an interactive component to it, it should never be executed.

FXschwartz replied:
@SandPod Done! I've moved the interactive configuration to be global. 

I have noticed a couple of potentially flaky tests that seem to be intermittently failing locally for me. `endpoint_method_test.dart`, `endpoint_file_test.dart`, and `endpoint_inheritance_analysis_test.dart` are failing sometimes with the error `PathNotFoundException: Deletion failed`. Might be a race condition with the temp directory cleanup?

SandPod reviewed (changes requested):
Nice clean PR! The code structure is easy to read and understand. Great work!

I have a couple of comments and suggestions to address I have left in the review, but then I would also like to to address three important things:
1. How does the user configure the interactive prompt?
2. Could you add an end to end test for this in the CLI e2e test directory: tests/serverpod_cli_e2e_test
3. Can we use the prompt tools from cli_tools instead of building a new one here?

FXschwartz commented on tools/serverpod_cli/lib/src/util/server_directory_finder.dart:
```diff
@@ -0,0 +1,314 @@
+/// 2. Searches child directories (depth 2)
+/// 3. Checks for sibling directories with standard naming patterns (only if not at boundary)
+/// 4. Searches upward through parent directories (max 5 levels or until boundary)
+/// 5. At each parent level, checks its children for server directories
```
@SandPod We "shouldn't" be no. I added a `visited` set that is used to manage already searched children. It gets passed to all `_searchChildrenRecursive` calls and there is a check at line 228 that if the current path exist in the normalized set we do an early return.

FXschwartz commented on tools/serverpod_cli/test/util/server_directory_finder_test.dart:
```diff
@@ -0,0 +1,494 @@
+    );
+
+    test(
+      'Given sibling directory with _flutter suffix, '
```
Makes a lot more sense. Updated!

SandPod commented on tools/serverpod_cli/lib/src/util/server_directory_finder.dart:
```diff
@@ -0,0 +1,314 @@
+    }
+
+    // 2. Search child directories (depth 2)
+    candidates.addAll(
```
I can't seem to find any tests that validate these early returns work as expected and in the given order that we have configured. Could you add a test that validates this?

SandPod commented:
> For your second point is adding a e2e test for the generate cli command enough, or do you also want tests for the create-migration

I think it is enough to test it with generate. We are only interested in whether the location algorithm works as expected e2e. Possibly also the `ci` override through command lines.

SandPod commented on tools/serverpod_cli/lib/src/config/config.dart:
```diff
@@ -238,7 +240,27 @@ class GeneratorConfig implements ModelLoadConfig {
+  /// The [interactive] parameter controls whether interactive prompts are enabled.
+  /// Defaults to true unless running in a CI environment (detected via ci package).
+  /// Explicit flag value overrides CI detection.
+  static Future<GeneratorConfig> load(
```
Suggestion: This is also called from `create` and `lsp`.

For the `create` command, the interactive flag should also be respected.
For the `lsp` command `interactive` should always be false (at least for now).

You can find them by searching for all references of this method.

Perhaps `interactive` should be nullable required parameter? So we ensure it is always handled.

FXschwartz replied:
Nice! I missed those. Updated both to respect the interactive flag with lsp being set to false, and made `interactive` nullable required.

SandPod commented on tools/serverpod_cli/lib/src/runner/serverpod_command_runner.dart:
```diff
@@ -125,6 +125,15 @@ enum GlobalOption<V> implements OptionDefinition<V> {
+    FlagOption(
+      argName: 'interactive',
+      negatable: true,
+      defaultsTo: true,
```
Suggestion: Remove default value.

Having a default value will always set this to true, resulting in it being impossible for us to detect if the value was set or not. By removing the default value, it will be set to `null` by default.

Add a test to `serverpod_command_runner_test.dart` that validates this behaves as expected.

FXschwartz commented:
@SandPod I believe all requested changes have been addressed and I've merged main and resolved conflicts. Take another looks when you have the bandwidth and let me know what you think!

FXschwartz commented:
@SandPod Ok I've manually ran dart format on those files and confirmed running `dart format --output=none --set-exit-if-changed .` locally passes. Fingers crossed we are good to go!

SandPod reviewed (approved):
LGTM! :rocket:
